### ISSUE TITLE
zfs-kmod - latest RHEL7.3 kernel not compatible

### ISSUE THREAD

**Issue by bziller** [zfsonlinux/zfs#5377]
The latest RHEL7.3 kernel 3.10.0-514.el7.x86_64 is not compatible with the kmod modules for 3.10.0-327.28.3.el7.x86_64:

```
# weak-modules --verbose --add-kernel 3.10.0-514.el7.x86_64
Module splat.ko from kernel 3.10.0-327.28.3.el7.x86_64 is not compatible   with kernel 3.10.0-514.el7.x86_64 in symbols: getrawmonotonic
Module spl.ko from kernel 3.10.0-327.28.3.el7.x86_64 is not compatible   with kernel 3.10.0-514.el7.x86_64 in symbols: getrawmonotonic fput out_of_line_wait_on_bit
Module zfs.ko from kernel 3.10.0-327.28.3.el7.x86_64 is not compatible   with kernel 3.10.0-514.el7.x86_64 in symbols: getrawmonotonic
```
```
# rpm -qa kmod-zfs kmod-spl
kmod-zfs-0.6.5.8-1.el7.centos.x86_64
kmod-spl-0.6.5.8-1.el7.centos.x86_64
```

**Comment by mmatuska** [comment]
I confirm this issue.

**Comment by ryao** [comment]
I hit this issue in the past 24 hours myself when working on ClusterHQ's secure boot capable RPM repository. I do not see an easy way of making this work. We are going without kapi-tracking packaging.

**Comment by behlendorf** [comment]
Unfortunately this was always a possibility because ZoL does use a few symbols not on the official white list.  For this to continue to work across **all** RHEL kernels we need to find a way to rework the existing code and remove the calls to the `getrawmonotonic`, `fput` and `out_of_line_wait_on_bit` functions.

As a stop gap the kmods for RHEL7 could be rebuilt against the new kernel.  They would then no longer work against old kernels but this is probably preferable to the current situation.

**Comment by ryao** [comment]
@behlendorf yum's dependency resolution might mess that up. You need to build against all kernels like I tried to do at the CHQ yum repositories:

```
sudo yum install https://d1cqii9i0878v1.cloudfront.net/el7/chq-release-1-1.el7.noarch.rpm
```

There is also one for Fedora 24:

```
sudo yum install https://d1cqii9i0878v1.cloudfront.net/fc24/chq-release-1-1.el7.noarch.rpm
```

People are welcome to use that. I will be maintaining it as part of my job. The url will change to yum.clusterhq.com as soon as I have https working. Automation is not in place yet, but it will be semi-automated soon. Full automation will be difficult because of security concerns involving the keys, but I am working on it. The packages and metadata is signed with a 4096-bit RSA key while the modules are signed with the private key for the following public key for secure boot compatibility:

https://d1cqii9i0878v1.cloudfront.net/chq-kmod.der

Also, I am the only person inside ClusterHQ with access to the package signing key and kernel module key at this time. That might change in the future, but my intention is to ensure that as few people have access to it as possible.

**Comment by kpande** [comment]
fwiw, we build our own CentOS kmod packages but we're not tracking upstream kernels.

**Comment by behlendorf** [comment]
Official packages have been published for RHEL/CentOS 7.3.

https://github.com/zfsonlinux/zfs/wiki/RHEL-&-CentOS#rhelcentos-73-kmod-packages
